sbt thinks that server is already booting because of this exception

## steps

Using sbt 1.6.1 on Windows 10. I am also using Amazon Corretto JDK 8.312, if that makes a difference. The issue is present in both Git Bash and cmd.

1. Open a terminal, cd to a project, and execute `sbt`.
2. Open another terminal, cd to the same project, and try executing `sbt` in any form, including `sbt --no-server`. This will give an error.

## problem

This is what I receive upon executing the second sbt command:
```
>sbt --no-server compile
sbt thinks that server is already booting because of this exception:
sbt.internal.ServerAlreadyBootingException: java.io.IOException: Could not create lock for \\.\pipe\sbt-load-8979016220299674466_lock, error 5
        at sbt.internal.BootServerSocket.newSocket(BootServerSocket.java:348)
        at sbt.internal.BootServerSocket.<init>(BootServerSocket.java:294)
        at sbt.xMain$.getSocketOrExit(Main.scala:145)
        at sbt.xMain$.bootServerSocket$lzycompute$1(Main.scala:77)
        at sbt.xMain$.bootServerSocket$1(Main.scala:77)
        at sbt.xMain$.withStreams$1(Main.scala:85)
        at sbt.xMain$.run(Main.scala:121)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at sbt.internal.XMainConfiguration.run(XMainConfiguration.java:57)
        at sbt.xMain.run(Main.scala:46)
        at xsbt.boot.Launch$.$anonfun$run$1(Launch.scala:149)
        at xsbt.boot.Launch$.withContextLoader(Launch.scala:176)
        at xsbt.boot.Launch$.run(Launch.scala:149)
        at xsbt.boot.Launch$.$anonfun$apply$1(Launch.scala:44)
        at xsbt.boot.Launch$.launch(Launch.scala:159)
        at xsbt.boot.Launch$.apply(Launch.scala:44)
        at xsbt.boot.Launch$.apply(Launch.scala:21)
        at xsbt.boot.Boot$.runImpl(Boot.scala:78)
        at xsbt.boot.Boot$.run(Boot.scala:73)
        at xsbt.boot.Boot$.main(Boot.scala:21)
        at xsbt.boot.Boot.main(Boot.scala)
Caused by: java.io.IOException: Could not create lock for \\.\pipe\sbt-load-8979016220299674466_lock, error 5
        at org.scalasbt.ipcsocket.Win32NamedPipeServerSocket.<init>(Win32NamedPipeServerSocket.java:129)
        at org.scalasbt.ipcsocket.Win32NamedPipeServerSocket.<init>(Win32NamedPipeServerSocket.java:48)
        at sbt.internal.BootServerSocket.newSocket(BootServerSocket.java:342)
        ... 23 more
Create a new server? y/n (default y)
```

## expectation

Ideally, the server would work without my intervention. If the server absolutely can't work, at least there should be a working way to disable it (again, without my intervention).

## notes

This issue, at least in some form, was reported previously in #6101, which has been marked as closed with #6728. However, the issue is still present for me.

This happens for me even on 1.8.2 on a vanilla, brand-new Ubuntu desktop install:

```
$ sbt
sbt thinks that server is already booting because of this exception:
sbt.internal.ServerAlreadyBootingException: java.io.IOException: org.scalasbt.ipcsocket.NativeErrorException: [13] Permission denied
	at sbt.internal.BootServerSocket.newSocket(BootServerSocket.java:356)
	at sbt.internal.BootServerSocket.<init>(BootServerSocket.java:295)
	at sbt.xMain$.getSocketOrExit(Main.scala:150)
	at sbt.xMain$.bootServerSocket$lzycompute$1(Main.scala:77)
	at sbt.xMain$.bootServerSocket$1(Main.scala:77)
	at sbt.xMain$.withStreams$1(Main.scala:85)
	at sbt.xMain$.run(Main.scala:121)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
	at sbt.internal.XMainConfiguration.run(XMainConfiguration.java:57)
	at sbt.xMain.run(Main.scala:46)
	at xsbt.boot.Launch$.$anonfun$run$1(Launch.scala:149)
	at xsbt.boot.Launch$.withContextLoader(Launch.scala:176)
	at xsbt.boot.Launch$.run(Launch.scala:149)
	at xsbt.boot.Launch$.$anonfun$apply$1(Launch.scala:44)
	at xsbt.boot.Launch$.launch(Launch.scala:159)
	at xsbt.boot.Launch$.apply(Launch.scala:44)
	at xsbt.boot.Launch$.apply(Launch.scala:21)
	at xsbt.boot.Boot$.runImpl(Boot.scala:78)
	at xsbt.boot.Boot$.run(Boot.scala:73)
	at xsbt.boot.Boot$.main(Boot.scala:21)
	at xsbt.boot.Boot.main(Boot.scala)
Caused by: java.io.IOException: org.scalasbt.ipcsocket.NativeErrorException: [13] Permission denied
	at org.scalasbt.ipcsocket.UnixDomainServerSocket.bind(UnixDomainServerSocket.java:140)
	at org.scalasbt.ipcsocket.UnixDomainServerSocket.<init>(UnixDomainServerSocket.java:114)
	at org.scalasbt.ipcsocket.UnixDomainServerSocket.<init>(UnixDomainServerSocket.java:93)
	at sbt.internal.BootServerSocket.newSocket(BootServerSocket.java:350)
	... 21 more
Caused by: org.scalasbt.ipcsocket.NativeErrorException: [13] Permission denied
	at org.scalasbt.ipcsocket.JNAUnixDomainSocketLibraryProvider.bind(UnixDomainSocketLibrary.java:178)
	at org.scalasbt.ipcsocket.UnixDomainServerSocket.bind(UnixDomainServerSocket.java:136)
	... 24 more
Create a new server? y/n (default y)
y
[info] welcome to sbt 1.8.2 (Private Build Java 19.0.1)
[info] loading settings for project ...```

--- a/main-command/src/main/java/sbt/internal/BootServerSocket.java
+++ b/main-command/src/main/java/sbt/internal/BootServerSocket.java
@@ -123,12 +123,17 @@ public class BootServerSocket implements AutoCloseable {
                             bytes.put(b);
                             clientSocketReads.put(ClientSocket.this);
                           } else {
+                            // close() deregisters from clientSockets like the write
+                            // methods do; a dead entry left behind would block the
+                            // NO_BOOT_CLIENTS_CONNECTED signal in inputStream.read.
                             alive.set(false);
+                            close();
                           }
                         }
 
                       } catch (IOException e) {
                         alive.set(false);
+                        close();
                       }
                     }
                   } catch (final Exception ex) {

--- a/main-command/src/main/scala/sbt/internal/BootServerSocketProbe.scala
+++ b/main-command/src/main/scala/sbt/internal/BootServerSocketProbe.scala
@@ -1,0 +1,42 @@
+/*
+ * sbt
+ * Copyright 2023, Scala center
+ * Copyright 2011 - 2022, Lightbend, Inc.
+ * Copyright 2008 - 2010, Mark Harrah
+ * Licensed under Apache License 2.0 (see LICENSE)
+ */
+
+package sbt.internal
+
+import java.util.concurrent.{ CountDownLatch, TimeUnit }
+import java.util.concurrent.atomic.AtomicBoolean
+import sbt.protocol.ClientSocket
+import scala.util.control.NonFatal
+
+private[sbt] object BootServerSocketProbe:
+  private val timeoutMillis = 2000L
+
+  /**
+   * True only if something answers on the boot socket at `location`. A live server answers
+   * immediately, so the connect runs on a daemon thread bounded by [[timeoutMillis]]: the
+   * underlying native connect has no timeout and blocks indefinitely against a bound socket whose
+   * listen backlog is saturated, which must never hang startup. LinkageError is caught alongside
+   * NonFatal because the connect may perform the JVM's first JNI/JNA load.
+   */
+  def liveServerDetected(location: String, useJni: Boolean): Boolean =
+    val answered = new AtomicBoolean(false)
+    val done = new CountDownLatch(1)
+    val t = new Thread(
+      () =>
+        try
+          ClientSocket.localSocket(location, useJni).close()
+          answered.set(true)
+        catch case NonFatal(_) | (_: LinkageError) => ()
+        finally done.countDown(),
+      "sbt-boot-socket-probe"
+    )
+    t.setDaemon(true)
+    t.start()
+    done.await(timeoutMillis, TimeUnit.MILLISECONDS)
+    answered.get()
+end BootServerSocketProbe

--- a/main-command/src/test/scala/sbt/internal/BootServerSocketSpec.scala
+++ b/main-command/src/test/scala/sbt/internal/BootServerSocketSpec.scala
@@ -1,0 +1,77 @@
+/*
+ * sbt
+ * Copyright 2023, Scala center
+ * Copyright 2011 - 2022, Lightbend, Inc.
+ * Copyright 2008 - 2010, Mark Harrah
+ * Licensed under Apache License 2.0 (see LICENSE)
+ */
+
+package sbt.internal
+
+import java.nio.file.{ Files, Paths }
+import sbt.internal.util.Util
+import verify.BasicTestSuite
+
+object BootServerSocketSpec extends BasicTestSuite:
+
+  // the constructor only reads baseDirectory; provider is never touched
+  private def config(base: java.io.File): xsbti.AppConfiguration =
+    new xsbti.AppConfiguration {
+      override def arguments(): Array[String] = Array.empty
+      override def baseDirectory(): java.io.File = base
+      override def provider(): xsbti.AppProvider = null
+    }
+
+  private def useJni: Boolean =
+    BootServerSocket.requiresJNI() || sys.props.getOrElse("sbt.ipcsocket.jni", "false") == "true"
+
+  private def probe(location: String): Boolean =
+    BootServerSocketProbe.liveServerDetected(location, useJni)
+
+  private def freshBase(prefix: String): (java.io.File, Long) =
+    val base = Files.createTempDirectory(prefix).toRealPath().toFile
+    (base, base.getAbsolutePath.hashCode.toLong ^ System.nanoTime())
+
+  test("a live boot server is detected by the probe") {
+    val (base, token) = freshBase("boot-socket-live")
+    val location = BootServerSocket.socketLocation(base.toPath, token)
+    val server = new BootServerSocket(config(base), token)
+    val live =
+      try probe(location)
+      finally server.close()
+    assert(live)
+  }
+
+  test("the probe reports no live server when nothing is listening") {
+    val (base, token) = freshBase("boot-socket-none")
+    val location = BootServerSocket.socketLocation(base.toPath, token)
+    val live = probe(location)
+    assert(!live)
+  }
+
+  test("after close, the probe reports no live server") {
+    val (base, token) = freshBase("boot-socket-closed")
+    val location = BootServerSocket.socketLocation(base.toPath, token)
+    val server = new BootServerSocket(config(base), token)
+    server.close()
+    val live = probe(location)
+    assert(!live)
+  }
+
+  test("a stale socket file is not a live server and does not block a new socket") {
+    if (!Util.isWindows) {
+      val (base, token) = freshBase("boot-socket-stale")
+      val location = Paths.get(BootServerSocket.socketLocation(base.toPath, token))
+      Files.createDirectories(location.getParent)
+      Files.createFile(location) // leftover from a killed process
+      val staleLooksLive = probe(location.toString)
+      assert(!staleLooksLive)
+      val server = new BootServerSocket(config(base), token) // reclaims the path
+      val liveAfterReclaim =
+        try probe(location.toString)
+        finally server.close()
+      assert(liveAfterReclaim)
+    }
+  }
+
+end BootServerSocketSpec

--- a/main/src/main/scala/sbt/Main.scala
+++ b/main/src/main/scala/sbt/Main.scala
@@ -152,29 +152,38 @@ private[sbt] object xMain:
   private def getSocketOrExit(
       configuration: xsbti.AppConfiguration
   ): (Option[BootServerSocket], Option[Exit]) = {
-    def printThrowable(e: Throwable): Unit = {
-      println("sbt thinks that server is already booting because of this exception:")
-      e.printStackTrace()
-    }
-
-    val target =
-      configuration.baseDirectory().toPath().toRealPath().resolve("project").resolve("target")
-    val hash = HashUtil.farmHash(target.toString().getBytes("UTF-8"));
+    val base = configuration.baseDirectory().toPath().toRealPath()
+    val target = base.resolve("project").resolve("target")
+    val hash = HashUtil.farmHash(target.toString().getBytes("UTF-8"))
+    // sbt runs fine without a boot socket (see the UnsatisfiedLinkError case below), so a
+    // creation failure only means "another sbt is booting" if something answers on it.
+    def liveServerDetected: Boolean =
+      BootServerSocketProbe.liveServerDetected(
+        BootServerSocket.socketLocation(base, hash),
+        BootServerSocket.requiresJNI() || SysProp.serverUseJni,
+      )
     try Some(new BootServerSocket(configuration, hash)) -> None
     catch {
-      case e: ServerAlreadyBootingException if hasConsole && !ITerminal.startedByRemoteClient =>
-        printThrowable(e)
+      // No live server and nothing the user can do about a socket failure, so proceed
+      // silently without the boot socket, as for UnsatisfiedLinkError below.
+      case _: ServerAlreadyBootingException if !liveServerDetected => (None, None)
+      case _: ServerAlreadyBootingException if hasConsole && !ITerminal.startedByRemoteClient =>
+        println("another sbt appears to be booting in this build.")
         println("Create a new server? y/n (default y)")
         val exit =
           if (ITerminal.get.withRawInput(System.in.read) == 'n'.toInt) Some(Exit(1))
           else None
         (None, exit)
-      case e: ServerAlreadyBootingException =>
+      case _: ServerAlreadyBootingException =>
         if (SysProp.forceServerStart) (None, None)
         else {
-          printThrowable(e)
+          println("another sbt appears to be booting in this build; exiting.")
+          println(
+            "wait for it to finish, attach to it with --client, or pass -Dsbt.server.forcestart=true to start anyway."
+          )
           (None, Some(Exit(2)))
         }
+      case _: IOException          => (None, None)
       case _: UnsatisfiedLinkError => (None, None)
     }
   }

--- a/notes/2.0.0/boot-socket-liveness.md
+++ b/notes/2.0.0/boot-socket-liveness.md
@@ -1,0 +1,16 @@
+### sbt no longer refuses to start when the boot socket cannot be created
+
+Failures to create the boot-time io socket used to be misreported: most were
+wrapped as "sbt thinks that server is already booting" with a stack trace, and
+non-interactive invocations exited with code 2, while failures to create the
+socket directory crashed startup outright. Permission or path-length problems
+with `XDG_RUNTIME_DIR` or the temp directory and Windows named-pipe access
+errors all hit one of these ([#6777][6777]).
+
+sbt now probes the socket first. Only a live server answering the probe is
+treated as another sbt booting in the build (the interactive prompt and
+non-interactive exit are unchanged, with a clearer message). Any other failure
+is no longer fatal: sbt continues without the boot socket, whose only job is
+forwarding boot-time io to early-connecting clients.
+
+  [6777]: https://github.com/sbt/sbt/issues/6777

